utils/log_parser.py

```diff
@@ -15,136 +15,214 @@
-        return self.colors[idx]
+    def get_color(self, node: Node) -> str:
+        """Returns the color for the given node. """
+
+        idx = hash(node.get_kind_name()) % len(self.colors_)
+        return self.colors_[idx]
+
+    def dump_node(self, node: Node) -> None:
+        """Generates the dotty information for the given node. """
 
-    def dump_node(self, node):
         if not node:
             return
-        nodeStr = self.get_unique_vertex_name(node) + "[\n"
-        nodeStr += "\tlabel = \""
-        nodeStr += self.dump_label(node)
-        nodeStr += "\"\n"
+        nodeStr = f"{self.get_unique_vertex_name(node)}[\n"
+        nodeStr += f"\tlabel = \"{self.dump_label(node)}\"\n"
         nodeStr += "\tshape = \"record\"\n"
         nodeStr += "\tstyle=\"filled,rounded\"\n"
-        nodeStr += "\tfillcolor="
-        nodeStr += self.get_color(node)
-        nodeStr += "\npenwidth = 2];\n"
+        nodeStr += f"\tfillcolor={self.get_color(node)}\n"
+        nodeStr += "penwidth = 2];\n"
         self.vertices_.append(nodeStr)
 
-    def visitNodes(self):
-        for k, n in NODES_MAP:
-            self.dump_node(NODES_MAP[(k, n)])
+    def visitNodes(self) -> None:
+        """Visits all nodes in NODES_MAP and dump the dotty information for each node. """
 
-    def visitEdges(self):
-        for k, n in NODES_MAP:
-            node = NODES_MAP[(k, n)]
+        for node in NODES_MAP.values():
+            self.dump_node(node)
+
+    def visitEdges(self) -> None:
+        """Visits all edges and dump the dotty information for each edge. """
+
+        for node in NODES_MAP.values():
             for nodeInput in node.get_inputs():
                 i = nodeInput[0]
                 if (i.get_kind_name(), i.get_name()) not in NODES_MAP:
                     print(i.get_kind_name(), i.get_name())
-                edgeStr = self.get_unique_vertex_name(i)+":Outputs -> "
-                edgeStr += self.get_unique_vertex_name(node)+":Inputs"
+                edgeStr = self.get_unique_vertex_name(i) + ":Outputs -> "
+                edgeStr += self.get_unique_vertex_name(node) + ":Inputs"
                 self.edges_.append(edgeStr)
 
-    def dump_graph(self, dagName):
+    def dump_graph(self, dagName: str) -> None:
+        """Visits the node graph and generates the dotty information. """
+
         self.visitNodes()
         self.visitEdges()
-        f = open(dagName + "_dotty.dot", "w")
-        f.write("digraph DAG {\n\trankdir=TB;\n")
-        for v in self.vertices_:
-            f.write(v + "\n")
-        for e in self.edges_:
-            f.write(e + ";\n")
-        f.write("}")
-        f.close()
+        with open(f"{dagName}_dotty.dot", "w") as f:
+            f.write("digraph DAG {\n\trankdir=TB;\n")
+            for v in self.vertices_:
+                f.write(v + "\n")
+            for e in self.edges_:
+                f.write(e + ";\n")
+            f.write("}")
+            f.close()
```

Comment:
`f.close()` will be called automatically at at the end of the `with` block so you don't need to do it explicitly.